smart-contract/program/src/processor/crank.rs

```diff
@@ -61,60 +62,85 @@ pub fn process_crank(
 ) -> ProgramResult {
     let accounts = Accounts::parse(accounts, program_id)?;
 
-    let present_time = Clock::get()?.unix_timestamp;
-
     let mut stake_pool = StakePool::get_checked(accounts.stake_pool, vec![Tag::StakePool])?;
-    let central_state = CentralState::from_account_info(accounts.central_state)?;
+    let mut central_state = CentralState::from_account_info(accounts.central_state)?;
+
+    let current_time = Clock::get()?.unix_timestamp;
+    let current_offset = (current_time - central_state.creation_time) / SECONDS_IN_DAY as i64;
+    // check if we need to do a system wide snapshot
+    if central_state.last_snapshot_offset < current_offset {
+        central_state.total_staked_snapshot = central_state.total_staked;
+        central_state.last_snapshot_offset = current_offset;
+        central_state.save(&mut accounts.central_state.data.borrow_mut())?;
+        // reset the delta for the pool that we are currently cranking. We don't want the history to influence our result
+        stake_pool.header.last_delta_update_offset = current_offset;
+        stake_pool.header.total_staked_delta = 0;
+    }
 
-    if present_time - stake_pool.header.last_crank_time < SECONDS_IN_DAY as i64 {
+    if stake_pool.header.current_day_idx as i64 == central_state.last_snapshot_offset {
         #[cfg(not(any(feature = "days-to-sec-10s", feature = "days-to-sec-15m")))]
         return Err(AccessError::NoOp.into());
     }
+
     msg!("Total staked in pool {}", stake_pool.header.total_staked);
     msg!("Daily inflation {}", central_state.daily_inflation);
     msg!("Total staked {}", central_state.total_staked);
+    msg!("Total staked delta {}", stake_pool.header.total_staked_delta);
+    msg!("Total staked snapshot {}", central_state.total_staked_snapshot);
+
+    // get the pool staked amount at the time of last system snapshot
+    let total_staked_snapshot: u128 = (stake_pool.header.total_staked - stake_pool.header.total_staked_delta as u64) as u128;
```

Comment:
You are casting a `i64` into a `u64` this is not safe

Reply:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=e00f0da11f00465a0fe161d1cfbf414f

Reply:
Good catch. Could you please check my solution? I am not sure if it is the right one, but I wanted to avoid the overflow of `stake_pool.header.total_staked as i64` as well

Reply:
You are still not solving the issue by casting everything into `i128` and then back into `u128` the issue is that casting signed integer into unsigned integer is unsafe.

Could do something like

```rust
    let total_staked_snapshot = match stake_pool.header.total_staked_delta.signum() {
        1 => stake_pool
            .header
            .total_staked
            .checked_sub(stake_pool.header.total_staked_delta.unsigned_abs()) // <- Actualy safe to cast as u64 here
            .ok_or(AccessError::Overflow)?,
        -1 => stake_pool
            .header
            .total_staked
            .checked_add(stake_pool.header.total_staked_delta.unsigned_abs())
            .ok_or(AccessError::Overflow)?,
        _ => stake_pool.header.total_staked,
    };
```

Reply:
Should also allow you to remove the unnecessary `if total_staked_snapshot > 0`

Reply:
Thank you for the proposed solution, I will use it.

I just want to doublecheck that my solution should never overflow as well, because we can safely assume that `stake_pool.header.total_staked as i128 - stake_pool.header.total_staked_delta as i128 > 0`. Or am I missing something?

As for the  `if total_staked_snapshot > 0`, I believe that we still need it, otherwise we would risk division by 0 on line
` .checked_div(central_state.total_staked_snapshot as u128)`

Reply:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=b1f261f46d4ccb75e91d798925d840ec

Reply:
The snippet that you have sent panics on division by 0 which is not something that we want here. We want to save the (0, 0) rewards pair into the rewards history